docs/codeql/codeql-language-guides/analyzing-data-flow-in-cpp-new.rst

```diff
@@ -0,0 +1,433 @@
+
+     override predicate isSink(DataFlow::Node sink) {
+       ...
+     }
+   }
+
+The following predicates are defined in the configuration:
+
+-  ``isSource``—defines where data may flow from
+-  ``isSink``—defines where data may flow to
+-  ``isBarrier``—optional, restricts the data flow
+-  ``isBarrierGuard``—optional, restricts the data flow
+-  ``isAdditionalFlowStep``—optional, adds additional flow steps
+
+The characteristic predicate ``MyDataFlowConfiguration()`` defines the name of the configuration, so ``"MyDataFlowConfiguration"`` should be replaced by the name of your class.
+
+The data flow analysis is performed using the predicate ``hasFlow(DataFlow::Node source, DataFlow::Node sink)``:
+
+.. code-block:: ql
+
+   from MyDataFlowConfiguration dataflow, DataFlow::Node source, DataFlow::Node sink
+   where dataflow.hasFlow(source, sink)
+   select source, "Data flow to $@.", sink, sink.toString()
+
+Using global taint tracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Global taint tracking is to global data flow as local taint tracking is to local data flow. That is, global taint tracking extends global data flow with additional non-value-preserving steps. The global taint tracking library is used by extending the class ``TaintTracking::Configuration`` as follows:
+
+.. code-block:: ql
+
+   import semmle.code.cpp.dataflow.new.TaintTracking
+
+   class MyTaintTrackingConfiguration extends TaintTracking::Configuration {
+     MyTaintTrackingConfiguration() { this = "MyTaintTrackingConfiguration" }
+
+     override predicate isSource(DataFlow::Node source) {
+       ...
+     }
+
+     override predicate isSink(DataFlow::Node sink) {
+       ...
+     }
+   }
+
+The following predicates are defined in the configuration:
+
+-  ``isSource``—defines where taint may flow from
+-  ``isSink``—defines where taint may flow to
+-  ``isSanitizer``—optional, restricts the taint flow
+-  ``isSanitizerGuard``—optional, restricts the taint flow
+-  ``isAdditionalTaintStep``—optional, adds additional taint steps
+
+Similar to global data flow, the characteristic predicate ``MyTaintTrackingConfiguration()`` defines the unique name of the configuration, so ``"MyTaintTrackingConfiguration"`` should be replaced by the name of your class.
+
+The taint tracking analysis is performed using the predicate ``hasFlow(DataFlow::Node source, DataFlow::Node sink)``.
+
+Examples
+~~~~~~~~
+
+The following data flow configuration tracks data flow from environment variables to opening files in a Unix-like environment:
+
+.. code-block:: ql
+
+   import cpp
+   import semmle.code.cpp.dataflow.new.DataFlow
+
+   class EnvironmentToFileConfiguration extends DataFlow::Configuration {
+     EnvironmentToFileConfiguration() { this = "EnvironmentToFileConfiguration" }
+
+     override predicate isSource(DataFlow::Node source) {
+       exists(Function getenv |
+         source.asIndirectExpr(1).(FunctionCall).getTarget() = getenv and
+         getenv.hasQualifiedName("getenv")
```

Comment:
It looks like `hasQualifiedName` is deprecated, in favour or `hasGlobalName`.

Reply:
Thanks. It's also missing the `deprecated` keyword. Which means this did not get flagged up anywhere.

Reply:
Using `hasGlobalName` now.